src/adapter/src/catalog/storage.rs

```diff
@@ -559,14 +555,50 @@ impl<S: Append> Connection<S> {
                 .value
                 + 1
         };
-        initialize_stash(&mut stash).await?;
-        migrate(&mut stash, skip, bootstrap_args).await?;
 
-        let conn = Connection {
+        // Initialize connection.
+        initialize_stash(&mut stash).await?;
+        let mut conn = Connection {
             stash,
             consolidations_tx,
         };
 
+        // Choose a time at which to apply migrations. This is usually the
+        // current system time, but with protection against backwards time
+        // jumps, even across restarts.
+        let initial_ts = {
+            let mut now_ts = now();
+            let initial_ts = conn
+                .try_get_persisted_timestamp(&Timeline::EpochMilliseconds)
+                .await?
+                .unwrap_or(mz_repr::Timestamp::MIN);
+            let initial_ts = cmp::max(initial_ts, now_ts.into());
+            let mut upper_bound = timeline::upper_bound(&mz_repr::Timestamp::from(now_ts));
+            while initial_ts > upper_bound {
+                // Cap retry time to 1s. In cases where the system clock has retreated by
+                // some large amount of time, this prevents against then waiting for that
+                // large amount of time in case the system clock then advances back to near
+                // what it was.
+                let remaining_ms = cmp::min(initial_ts.saturating_sub(upper_bound), 1_000.into());
+                error!(
+                    "Coordinator tried to start with initial timestamp of \
+                    {initial_ts}, which is more than \
+                    {TIMESTAMP_INTERVAL_UPPER_BOUND} intervals of size {} larger than \
+                    now, {now_ts}. Sleeping for {remaining_ms} ms.",
+                    *TIMESTAMP_PERSIST_INTERVAL
+                );
+                thread::sleep(Duration::from_millis(remaining_ms.into()));
+                now_ts = now();
+                upper_bound = timeline::upper_bound(&mz_repr::Timestamp::from(now_ts));
+            }
+            // IMPORTANT: we durably record the new timestamp before using it.
+            conn.persist_timestamp(&Timeline::EpochMilliseconds, initial_ts)
+                .await?;
+            initial_ts
+        };
```

Comment:
I think that this will work, but it does feel like a bit of an abstraction break that we do this outside of the context of the Timestamp Oracle. Then later when we initialize the Timestamp Oracle we're relying on this having already been done earlier.

Reply:
Then again, pushing the timestamp up the the current system time or waiting for the system time to catch up to the timestamp oracle isn't strictly necessary for correctness. So the Timestamp Oracle isn't relying on this happening for correctness, just for correlation to the system clock.

It just feels a little weird to update the durable timestamp outside of the context of the Timestamp Oracle.

Reply:
> OTOH it's not like this code was inside `DurableTimestampOracle` before?

Yeah but I think that was already an abstraction break and it should have been inside of `DurableTimestampOracle`. I think moving it to a function inside of `timeline.rs` is a good compromise.

Reply:
Ok, done. Lemme know what ya think.